Exception/InternalError.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Hoa\Compiler\Exception;
+
+use LogicException;
+
+/**
+ * It probably points to some internal issue of the Hoa Compiler library.
+ * Regardless source of the bug, try to report about this exception to the library maintainers.
+ * Even if bug is yours, this exception must not happen.
+ */
+final class InternalError extends LogicException
```

Comment:
`final` is not required here. I think you can remove it.

Reply:
All exceptions must extend `Hoa\Exception`, see sibling exception in the same directory.

Reply:
I would also name the class `RegularExpression` or `PCRE`, something less generic than `InternalError`.

Reply:
> final is not required here. I think you can remove it.

Let's try to see it from another point of view: there is no `final` keyword. There is `open` one. You suggest to open for inheritance. Why? I mean, it's known idea that composition is preferable over inheritance for some objective reasons. So it could be just a language design mistake to make classes open to inheritance by default. Some modern languages, for example, [fixed it](https://kotlinlang.org/docs/reference/classes.html). So what's the real reason to ~remove `final`~ open class for inheritance? 

> I would also name the class RegularExpression or PCRE, something less generic than InternalError.

I don't get it either: `RegularExpression` is a special language or pattern, depends on context, but not an error. Kinda weird for exception class name, IMO. You probably rely on namespaces, but I 

I understand and respect your own coding style, but I'm curious how do you explain it.

Reply:
We don't use `final` in the existing code base. I prefer to address that as another PR if you don't mind :-). You are pointing to an interesting approach, and I like it, but yeah, keep things seperate a little bit :-).

About the class/exception name, maybe `REgularExpression` is not appropriate. But I don't find `InternalError` more appropriate, it's too much abstract.

The exception represents an error in the lexer. We can (i) reuse the `Hoa\Compiler\Exception\Lexer` exception class (https://github.com/hoaproject/Compiler/blob/master/Exception/Lexer.php), or (ii) create another one to reflect more precisely what is the origin of the exception, hence a better name than `InternalError`.

If you decide to re-use `Exception\Lexer`, I'm fine with that. I would even suggest to do that.

Reply:
`Hoa\Compiler\Exception\Lexer` doesn't look like internal error of the `Lexer`. In other words, internal error is library's failure, user did nothing wrong (at least how I interpret it; that's why I reused it for `Text is not valid utf-8 string, you probably need to switch "lexer.unicode" setting off.`, obvious misconfiguration by user). It's like comparing 4xx HTTP error codes to 5xx.

I don't like idea to mix up semantically different errors into one exception. 

`InternalError` is definitely generic name, I consider this exception as unchecked one; users shouldn't be aware of this exception at all. Thus, they shouldn't catch this type of exception directly (only like `\Exception` or `\Throwable` in specific parts of application where failure of subprogram can be handled). 

The thing either works or it is broken. Classifying hypothetical errors is a waste of time for me as most of them are really hypothetical: you don't expect them, otherwise you'd just fixed the; you just make some "save point" for yourself to make sure that it works internally like you expected without surprises. It is similar to `Assertion::blahBlah()`: you don't care about exception class. Assertion just must not fail. If it does, your program is bugged like if you passed object as argument to integer parameter. 

And regarding names. Practice says that verbs are way more important in programming than nouns. Stack trace says `Lexer` is broken and `Lexer` throws `Lexer` exception. I want to know what _happened_ wrong. `LexerEncounteredInvalidUtf8Sequence`, `TokenIsEmptyString`, etc.

I also was confused when I found class `Lexer` in unit-tests which tried to mimic test of real `Lexer`. :)

But I can reuse `Hoa\Compiler\Exception\Lexer`.